storage/backup/backup-cli/src/coordinators/restore.rs

```diff
@@ -88,110 +89,284 @@ impl RestoreCoordinator {
+    /// We basically introduce a ledger history start version (lhs), a replay start version (rs) and target version
+    /// We directly store the write set and key values between (lhs, rs) and replay txn from (rs, target]
+    async fn run_impl(self) -> Result<()> {
         if self.replay_all {
             bail!("--replay--all not supported in this version.");
         }
 
+        info!("This tool only guarantees resume from previous in-progress restore. \
+        If you want to restore a new DB, please either specify a new target db dir or delete previous in-progress DB in the target db dir.");
+
         let metadata_view = metadata::cache::sync_and_load(
             &self.metadata_cache_opt,
             Arc::clone(&self.storage),
             self.global_opt.concurrent_downloads,
         )
         .await?;
 
-        let next_txn_version = self
+        let target_version = self.global_opt.target_version;
+        COORDINATOR_TARGET_VERSION.set(target_version as i64);
+
+        // calculate the start_version and replay_version
+        let max_txn_ver = metadata_view
+            .max_transaction_version()?
+            .ok_or_else(|| anyhow!("No transaction backup found."))?;
+        let lhs = self.ledger_history_start_version();
+
+        let latest_tree_version = self
+            .global_opt
+            .run_mode
+            .get_state_snapshot_before(Version::MAX);
+        let tree_completed = {
+            match latest_tree_version {
+                Some((ver, _)) => self
+                    .global_opt
+                    .run_mode
+                    .get_state_snapshot_before(ver)
+                    .is_some(),
+                None => false,
+            }
+        };
+
+        let mut db_next_version = self
             .global_opt
             .run_mode
             .get_next_expected_transaction_version()?;
-        if next_txn_version != 0 {
-            // DB is already in workable state
+
+        let kv_snapshot = if db_next_version == 0 {
+            match self.global_opt.run_mode.get_in_progress_state_kv_snapshot() {
+                Ok(Some(ver)) => {
+                    let snapshot = metadata_view.select_state_snapshot(ver)?;
+                    ensure!(
+                        snapshot.is_some() && snapshot.as_ref().unwrap().version == ver,
+                        "cannot find in-progress state snapshot {}",
+                        ver
+                    );
+                    snapshot
+                },
+                Ok(None) | Err(_) => {
+                    metadata_view.select_state_snapshot(std::cmp::min(lhs, max_txn_ver))?
+                },
+            }
+        } else {
+            None
+        };
+
+        let tree_snapshot = metadata_view
+            .select_state_snapshot(std::cmp::min(self.target_version(), max_txn_ver))?
+            .expect("Cannot find tree snapshot before target version");
```

Comment:
when tree is partial (exists but not complete), we need to get the exact snapshot to resume, similar to what you did for the kv snapshot